[Tooling] Migrate download_metadata.swift to gp_downloadmetadata action

AliSoftware opened the pull request:
This work is part of paaHJt-2Ib-p2 — task [i2]

⚠️  This PR targets the `release/19.7` branch, as this is where it's the most likely that the changes from this one and the relevant lanes will be tested next.

### Why / How

As an effort to migrate all our legacy Localization tooling and scripts into our release-toolkit fastlane plugin, this PR:
 - Removes the `fastlane/download_metadata.swift` script and replace it with a call to our existing `gp_downloadmetadata` action
 - Moves the metadata files that were in `fastlane/appstoreres/{metadata,jetpack_metadata}/source/{description,keywords,name,subtitle}.txt` into `fastlane/{metadata,jetpack_metadata}/default` instead, for consistency.
 - Removes the `fastlane/{metadata,jetpack_metadata}/en-US/{description,keywords,name,subtitle,release_notes}.txt` files (and add a `.gitkeep` file in `en-US` dir to still keep it around) so that `default/` is the new source of truth (used both for `en-US` and for any other locales that would not have translations available in GlotPress for those keys)
   - The `default/` pseudo-locale folder is now the source of truth for the originals of the metadata copies that we translate and include in our translation pipeline and automation, and in theory should act as a fallback for locales missing translations(†).
   - Some `.txt` files are still present in `en-US` for now though. Those are the files that haven't been configured to be included in our translation pipeline yet (files that were not treated by `download_metadata.swift` and not listed in `APPSTORECONNECT_METADATA_KEYS_MAP` yet). Those copies (`apple_tv_privacy_policy`, `marketting_url`, …) change rarely (if ever?), and in the current state of the repo all seem duplicated in the other locale folders (and our script never touch those — they seem to always been manually updated apparently). In a future PR, we might consider either adding those extra files to the pipeline (i.e. to the `AppStoreStrings.po` file / `ios_update_metadata_source` call, and in the `APPSTORECONNECT_METADATA_KEYS_MAP` map)… or keeping them untranslated and out of our translation pipeline/automation — but in that case we ought to delete all duplicate `.txt` files with English copy from each locale folder, and move those `.txt` files to `default/` for a single source of truth. All this seems outside of the scope of this PR though, so I preferred to keep that for later to keep this PR focused on porting the existing configuration only.
 - Make the lanes copy the `WordPress/{Jetpack,}/Resources/release_notes.txt` original copy into 
`fastlane/{metadata,jetpack_metadata}/default/release_notes.txt` at the same time that it download the translations for other locales from GlotPress, so that the Editorial Copy for the release notes is available in the path expected by `deliver`
    - In the future we hope to make that `fastlane/{metadata,jetpack_metadata}/default/release_notes.txt` path the official source of truth that we'll use as part of our Release Scenarios and the automation for the Announcement P2 post etc. But this is work for a future iteration that felt outside of the scope of this project.

## To Test

 - 🕵️  Check the code logic and the map of locales matches what was in the deleted `download_metadata.swift` — to ensure we still translate the same locales
 - ✂️  Cut a test branch from this one
 - 😈  Remove some metadata folders like `fastlane/jetpack_metadata/fr` and/or `fastlane/metadata/da` (so that we can check that they'll be re-created, even if empty) and commit.
 - 💻  Run `bundle exec fastlane download_wordpress_localized_app_store_metadata` then `bundle exec fastlane download_jetpack_localized_app_store_metadata` 
 - 🕵️  Check that the action worked and created 2 commits, one for WordPress and one for Jetpack
 - 🕵️  Check the content of the commits:
   - The `{description,keywords,name,subtitle}.txt` files in both `fastlane/metadata/*/` and `fastlane/metadata/*/` were unchanged (modulo CRLF at end of file which can be ignored) — since it's unlikely that new translations for those came in since last time we updated them.
   - The folders for locales you deleted in the 😈  step got recreated and re-populated with the proper copies
   - The copies for locales which didn't have any translations in GlotPress (e.g. `da`, `en-AU`, …) got deleted(‡), but their corresponding folder still got created with a `.gitkeep` file, even if no translation / `*.txt` files were found for that locale
   - The `release_notes.txt` files got copied from `WordPress/{Jetpack,}/Resources/release_notes.txt` into `fastlane/{metadata,jetpack_metadata}/default/` for WP and JP
 - 🧹 Delete your test branch

## Footnotes

_(†) It will probably be worth checking during the next AppStore submission that this indeed works as expected with `deliver` — especially check locales missing some translations for (some of all of) their metadata still got updated, but with the fallback copy from `default/`. Especially thinking about cases like `metadata/da/name.txt` which was still `"WordPress"` from a previous original copy while the new copy was never translated; with the changes from this PR, `deliever` should update the `da` locale in ASC and fallback to `metadata/default/name.txt` (aka `"WordPress – Website Builder"`) for the copy used._

_(‡) This shows that this PR fixed a bug where the previous `download_metadata.swift` failed to delete copies missing a translation… which led to old copies being used if a translation for new ones were not available in GlotPress… while the new setup ensures the folder exists but is empty for those cases, which should make `deliver` still update those locales but fallback to the copies in `default/` for those._

## What's next

Once a similar change has been adopted in all client apps (i.e. being ported WCiOS if relevant), we should consider removing the legacy methods relying on that script in the release-toolkit:

 - remove the `skip_glotpress` parameter (and associated code using it) in `ios_bump_version_release` logic to stop doing the updates at all (to do the same behavior as if the parameter was always `false`)
 - remove the `include_metadata:` parameter from `Ios::GitHelper::commit_version_bump` as well

We could at least mark them as explicitly deprecated in the next minor release, and remove them in the next major release.

AliSoftware committed:
Move metadata source files for consistency

Moved `description.txt`, `keywords.txt`, `name.txt` and `subtitle.txt` from `fastlane/appstoreres/*metadata/source` to `fastlane/*metadata/default` for consistency as a source of truth.

Note that `release_notes.txt` will be copied to `fastlane/*metadata/default` from `WordPress/Resources/release_notes.txt` on each sprint during the `download_*_localized_app_store_metadata` lanes (until we make that location the new source of truth for release notes later too)

AliSoftware committed:
`download_metadata.swift` -> `gp_downloadmetadata`

AliSoftware committed:
Avoid duplicating `default/*.txt` files in `en-US`

AliSoftware committed:
Reorder JP files list for consistency with WP

AliSoftware committed:
Stop trying to update `download_metadata.swift` on version bumps

`ios_bump_version_release` would otherwise try to update the `let glotPressWhatsNewKey` constant from `download_metadata.swift` file when bumping the release version.

AliSoftware committed:
Fix metadata_directory variable definition

AliSoftware committed:
Provide separate list of locales for WP vs JP

AliSoftware committed:
Ensure .gitkeep files exist in each locale's folder

To prevent a locale folder which we have listed in our Fastfile for WordPress/Jetpack to be ignored by `deliver` if it ends up empty because no translation is available in GlotPress

AliSoftware committed:
Add `.gitkeep` files to existing locale folders

mokagio commented on fastlane/lanes/localization.rb:
```diff
@@ -223,50 +266,80 @@
+    metadata_directory = File.join(PROJECT_ROOT_FOLDER, 'fastlane', 'metadata')
 
-    # @TODO: Make the `fastlane/metadata/en-US/release_notes.txt` path be the source of truth for the original copies in the future.
+    # FIXME: We should make the `fastlane/metadata/default/release_notes.txt` path be the source of truth for the original copies in the future.
```
I'm gathering we're dropping the `@` prefix for `TODO` and `FIXME` comments?

I never looked into where the convention came from. I don't mind either way. Just asking to hopefully remember next time I find myself writing such notes 😄

AliSoftware replied:
Yeah I think I have thought/assumed for a long time that the `@` prefix was expected and only recently discovered that rubocop and RubyMine suggest just `FIXME` and `TODO` without any `@` so I think it must have been just me adding that prefix for no reason after all 😅

AliSoftware replied:
Refs:
 - https://docs.rubocop.org/rubocop/1.27/cops_style.html#stylecommentannotation
 - https://rubystyle.guide/#annotate-keywords

mokagio commented on fastlane/lanes/localization.rb:
```diff
@@ -223,50 +266,80 @@
+      locales: locales_map,
+      download_path: metadata_directory
+    )
+    files_to_commit = [File.join(metadata_directory, '**', '*')]
```
Any particular reason for dropping the `.txt` extension what was in the previous version (at line `-265`)?

Just curious. I think it's fine. Realistically, there shouldn't be other files in those locations.

AliSoftware replied:
Good catch; it's a leftover from previous attempts at using `*` to include `.gitkeep` files as well, until I realized after testing it that `*` ignores dotfiles so that didn't work.

Now that I've moved to the new approach with `files_to_commit.append(…)` to cover the case of `.gitkeep` files, we could restore the `*.txt` here from before indeed.

AliSoftware replied:
```suggestion
    files_to_commit = [File.join(metadata_directory, '**', '*.txt')]
```

mokagio commented on fastlane/lanes/localization.rb:
```diff
@@ -223,50 +266,80 @@
+    )
+    files_to_commit = [File.join(metadata_directory, '**', '*')]
+
+    # Ensure that none of the `.txt` files in `en-US` would accidentally override our originals in `default` by mistake
+    APPSTORECONNECT_METADATA_KEYS_TO_FILES_MAP.values.map { |h| h[:desc] }.each do |file|
+      en_file_path = File.join(metadata_directory, 'en-US', file)
+      UI.user_error!("File `#{en_file_path}` would override the same one in `#{metadata_directory}/default`, but `default/` is our source of truth for those. " \
```
Minor suggestion that, IMHO, makes the comment and error slightly more concise.

```suggestion
    # Ensure that none of the `.txt` files in `en-US` would accidentally override our originals in `default`
    APPSTORECONNECT_METADATA_KEYS_TO_FILES_MAP.values.map { |h| h[:desc] }.each do |file|
      en_file_path = File.join(metadata_directory, 'en-US', file)
      UI.user_error!("File `#{en_file_path}` would override the same one in `#{metadata_directory}/default`, but `default/` is the source of truth. " \
```

mokagio commented on fastlane/lanes/localization.rb:
```diff
@@ -223,50 +266,80 @@
-    release_notes_source = File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Jetpack', 'Resources', 'release_notes.txt')
-    FileUtils.cp(release_notes_source, File.join(metadata_directory, 'en-US', 'release_notes.txt'))
+    # In the meantime, just copy the file to the right place for `deliver` to find, for the `default` pseudo-locale which is used as fallback
+    release_notes_source = File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Resources', 'release_notes.txt')
```
I think this should be 

```suggestion
    release_notes_source = File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Jetpack', 'Resources', 'release_notes.txt')
```

It's easy to miss because for the past few releases we had the same notes between the two apps, but there's always a chance they'll diverge in a future release.

AliSoftware replied:
Ah! The funny thing is that:
- One of the reasons I intentionally put that code in each of the lanes (as opposed to putting this logic in the `def download_localized_appstore_metadata` helper) was exactly because the paths for the source file were different in the 2 cases… but then I apparently did one too many copy/paste and forgot to finally make them distinct 🤦 
 - I had in mind to add a specific step in the Testing Instructions to mention "Edit the `release_notes.txt` files mentioned at the path in the Release Scenario to provide different copies for the WP and JP files before running the lane, then check the right ones were copied in `default` subfolder of the right metadata folders"… but then I forgot about it when actually writing the PR description 😅

So nice one catching this one despite it being sneaky here 😀👏👍

AliSoftware replied:
Actually just realized that I _did_ use different `release_notes_source` for WP and JP. It's just that… I swapped them 😓  So I should also apply [this other fix on the other side](https://github.com/wordpress-mobile/WordPress-iOS/pull/18439#discussion_r858167044) 😅

mokagio reviewed (changes requested):
Tested with https://github.com/wordpress-mobile/WordPress-iOS/pull/18441/ and everything looked good.

I'm requesting changes because of the Jetpack release notes source of truth path, but other than that, this would be good to go.

If you find yourself slammed with work and don't get to this, I can always apply the change and merge before finalizing 19.7 in a few days.

AliSoftware commented on fastlane/lanes/localization.rb:
```diff
@@ -223,50 +266,80 @@
-      message: 'Update WordPress metadata translations',
-      allow_nothing_to_commit: true
+    # In the meantime, just copy the file to the right place for `deliver` to find, for the `default` pseudo-locale which is used as fallback
+    release_notes_source = File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Jetpack', 'Resources', 'release_notes.txt')
```
```suggestion
    release_notes_source = File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Resources', 'release_notes.txt')
```

AliSoftware committed:
Fix paths for WP and JP release notes files being swapped

h/t @mokagio for noticing!

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

AliSoftware committed:
Improve comment and error wording

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

AliSoftware committed:
Restore proper glob for `*.txt` metadata files to commit

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

AliSoftware commented on fastlane/lanes/localization.rb:
```diff
@@ -223,50 +266,80 @@
-    FileUtils.cp(release_notes_source, File.join(metadata_directory, 'en-US', 'release_notes.txt'))
-
-    metadata_path = File.join(metadata_directory, '**', '*.txt')
-    git_add(path: metadata_path, shell_escape: true)
```
@mokagio  ⚠️ Notice how you used `shell_escape: true` before, but I changed it to `shell_escape: false` [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/18439/files#diff-ad77c85d7eb3034dec610211cb1747f7cac134e425448b1dcc246d71df3e8639R339).

I think `shell_escape: false` is the right choice, because we pass a parameter looking like `fastlane/metadata/**/*.txt` so we **want** this to be passed **unescaped** to the `git add` shell command, so that the shell interprets those `*` as glob patterns, and matches them all. While `shell_escape: true` would lead to `git_commit` adding shell escapes to the paths and transforming them to `fastlane/metadata/\*\*/\*.txt` and trying to match files with literal `*` characters in their name.

That being said, I'm a bit surprised that `git_add(path: metadata_path, shell_escape: true)` worked before that PR. I think we were just lucky because that `git_add` would then not find any file with those literal `*` characters and thus add nothing (`git add` would be a no-op there… instead of erroring?!), and then `git_commit` would take care of doing the staging and adding of the files to git (which only works with files that were already known to git and wouldn't work for brand new files, but we were probably lucky to never encounter that use case in practice, since that code was introduced recently?)

mokagio replied:
> That being said, I'm a bit surprised that `git_add(path: metadata_path, shell_escape: true)` worked before that PR. I think we were just lucky because that `git_add` would then not find any file with those literal `*` characters and thus add nothing (`git add` would be a no-op there… instead of erroring?!), and then `git_commit` would take care of doing the staging and adding of the files to git (which only works with files that were already known to git and wouldn't work for brand new files, but we were probably lucky to never encounter that use case in practice, since that code was introduced recently?)

🤔 

I added that `git_add(path: ..., shell_escape: true)` via 18f3bf73d1314f58321f3236a3a6a3aabef5be2b, but I can't remember why exactly. I added good details in the commit message about why the change was necessary, but not about why I went out of my way to explicitly pass the `shell_escape` option, set to the default value 🤔 

However, `git add` with `shell_escape: true` is _not_ a no-op. It indeed adds any file matching the pattern... even if escaped 🤔 

<img width="979" alt="image" src="https://user-images.githubusercontent.com/1218433/165236300-40f18eea-ca94-441b-968f-cea40e62c16d.png">

But, that might be because of Git itself:

<img width="850" alt="image" src="https://user-images.githubusercontent.com/1218433/165236443-754cf33c-b2ea-43a3-8cd7-b81d402c0986.png">

I'm running out of time in my day to explore this in more details. I'm happy with `false`, which is after all what the documentation recommends to use in this case:

```
+--------------+------------------------+------------+---------+
|                       git_add Options                        |
+--------------+------------------------+------------+---------+
| Key          | Description            | Env Var(s) | Default |
+--------------+------------------------+------------+---------+
| path         | The file(s) and        |            |         |
|              | path(s) you want to    |            |         |
|              | add                    |            |         |
| shell_escape | Shell escapes paths    |            | true    |
|              | (set to false if       |            |         |
|              | using wildcards or     |            |         |
|              | manually escaping      |            |         |
|              | spaces in :path)       |            |         |
```

(Maybe that's what happened! I read the docs, saw we'd need the option, but used the wrong value somehow 🤷‍♂️ )

AliSoftware reviewed (commented):
Adding a note to highlight a change on `shell_escape` you might have missed, for confirmation 🙂

AliSoftware commented:
@mokagio I think I've addressed all your suggestions.
Feel free to re-review (e.g. test with dummy release notes with different copies for WP vs JP to validate the right files are copies, this time?) — including checking the case about shell escape I mentioned above — then merge the PR yourself if all looks good to you then.

---

PS: Be sure to remember to double-check the metadata this Friday during submission to AppStore — especially for the cases like `fastlane/metadata/da/name.txt` I mentioned in the PR description, to ensure it falls back to the `default/` copy rather than not updating it at all and/or still using the translation from the previous copy (given the new copy doesn't have a `da` translation in GotPress so far), etc. 
_Could be worth manually adding an item on the Release Scenario for 19.7 to not forget?_

AliSoftware commented on fastlane/lanes/localization.rb:
```diff
@@ -223,50 +266,80 @@
+    # FIXME: Replace this with a call to the future replacement of `gp_downloadmetadata` once it's implemented in the release-toolkit (see paaHJt-31O-p2).
+
+    locales_map = GLOTPRESS_TO_ASC_METADATA_LOCALE_CODES.slice(*locales)    
+    what_new_key = "v#{ios_get_app_version}-whats-new"
+    gp_downloadmetadata(
+      project_url: glotpress_project_url,
+      target_files: APPSTORECONNECT_METADATA_KEYS_TO_FILES_MAP,
```
😱  Whoops, just noticed an issue there (introduced after refactoring the list of target files into a constant during my work):

 - `what_new_key` (which should be `whats_new_key`, really, eh) is defined on line 313 but never used
 - `APPSTORECONNECT_METADATA_KEYS_TO_FILES_MAP` is now a constant, which [uses a hardcoded `what_new_key` key (verbatim)](https://github.com/wordpress-mobile/WordPress-iOS/pull/18439/files#diff-ad77c85d7eb3034dec610211cb1747f7cac134e425448b1dcc246d71df3e8639R90), instead of relying on the variable defined on line 313 here to know which key to use.


While we've talked (in our respective RFCs for the refactor of the `.po` files and metadata handling) about unifying the key for the release notes, and could even consider using a static, invariant key name (rather than one that contains the version in its key) in the future… this is outside of the scope of this PR (and would fail to find the release notes for 19.7 anyway, given translations for those use the `v19.7-whats-new` key, not a key like `what_new_key`…)

We'll have to fix this before merging.

AliSoftware replied:
Should be fixed via b61cfeda85891554910f43a69812ae3eb0b60bb7

I've tried by running `download_wordpress_localized_app_store_metadata` on top of this fix and indeed I'm getting a couple of translated `release_notes.txt` files now (at least the ones already translated in GlotPress) 🎉

AliSoftware committed:
Use correct "what's new" key in `keys=>files` map

mokagio reviewed (approved):
👍 

Confirming the release notes fetching changes from [`b61cfed`](https://github.com/wordpress-mobile/WordPress-iOS/pull/18439/commits/b61cfeda85891554910f43a69812ae3eb0b60bb7) worked. Kudos on catching them. Phew, I'm glad you did, else I would have had to scramble during release finalization 😅

mokagio commented:
> Could be worth manually adding an item on the Release Scenario for 19.7

Yeah, fair enough

<img width="796" alt="image" src="https://user-images.githubusercontent.com/1218433/165238587-b768b970-c1b6-4299-8ac9-4dc6c9f87668.png">

mokagio commented:
> PS: Be sure to remember to double-check the metadata this Friday during submission to AppStore — especially for the cases like `fastlane/metadata/da/name.txt` I mentioned in the PR description, to ensure it falls back to the default/ copy rather than not updating it at all and/or still using the translation from the previous copy (given the new copy doesn't have a da translation in GotPress so far), etc.

This is what `deliver` proposed to upload. As expected, it uses the value from `default` (still... 🎉 ).

<img width="1487" alt="image" src="https://user-images.githubusercontent.com/1218433/165887905-59951d49-572c-4ad8-aacf-ee14e5c744b1.png">

mokagio commented:
When submitting Jetpack, I got this alert, which confirms ASC received the Jetpack `name.txt` value.

<img width="1017" alt="image" src="https://user-images.githubusercontent.com/1218433/165890078-f7f9751e-a119-4fa9-b8c7-f0464fdcd080.png">

That actually made me realize we don't have the name up for localization 🤔 That's something I'll need to check with the team and maybe address next release.

mokagio commented:
And this is the WordPress notice:

<img width="1033" alt="image" src="https://user-images.githubusercontent.com/1218433/165893807-0d815254-e441-4107-807b-1e1dcb24ef1b.png">

---

I was a bit unsure about using the full ["WordPress – Website Builder"](https://github.com/wordpress-mobile/WordPress-iOS/blob/release/19.7/fastlane/metadata/default/name.txt) name in non-English locales, as opposed to the current "WordPress".

I can see the benefits both ways:

- "WordPress" is the brand name, so it's not surprising that is not translated (e.g. in Italian it would be "PressaParola" or something like that 🙃 )
- "WordPress – Website Builder" might look odd to a non-English speaker, but it does add those two keywords, which the App Store might be using

⚖️ 

It's easy enough to change for both WordPress and Jetpack, so even if 19.7 might be going out with a non-ideal configuration, we'll be able to change it in two weeks time.

AliSoftware commented:
Thanks for checking and reporting back!
